2.6.x unsafe if-modified-since handling (breaks assets)

### Play Version (2.5.x / etc)
2.6.3


### API (Scala / Java / Neither / Both)
Scala


### Operating System (Ubuntu 15.10 / MacOS 10.10 / Windows 10)

Linux 4.13

### JDK (Oracle 1.8.0_72, OpenJDK 1.8.x, Azul Zing)

hotspot 1.8.0_144

### Expected Behavior

On any play application with an assets route (I'm using `GET           /assets/*file                             controllers.Assets.versioned(path="/public/assets", file: Asset)` as route here)

```bash
curl 'http://localhost:9000/assets/test' -H 'If-Modified-Since: Mon, 26 Jul 1997 05:00:00 GMT' 
```

should work.

I found this out the hard way because my firefox browser sends this header for specific assets.
UPDATE: This seems to be some stupid javascript that sets a global if-modified-since header that is not necessarily correct.
Still I would expect play to handle this gracefully (maybe with a 4xx status code but ideally just serve the asset anyway)

### Actual Behavior
The entire play asset stack explodes.
The date parsing code seems to be throwing away the time zone and thus ends up in a illegal state.

```
[error] a.m.f.ErrorHandler - flow_id=7a7b5f52-92d9-11e7-bfcc-9cb6d0d9871d, request_uri=/assets/shapes/germany.geojson - Internal server error GET /assets/shapes/germany.geojson             
java.lang.RuntimeException: Unexpected error while serving shapes/germany.geojson at /public/assets: Text 'Mon, 26 Jul 1997 05:00:00' could not be parsed: Conflict found: Field DayOfWeek 6 differs from DayOfWeek 1 derived from 1997-07-26
        at controllers.AssetsBuilder$$anonfun$assetAt$6.applyOrElse(Assets.scala:760)         
        at controllers.AssetsBuilder$$anonfun$assetAt$6.applyOrElse(Assets.scala:755)         
        at scala.concurrent.Future.$anonfun$recoverWith$1(Future.scala:412)                   
        at scala.concurrent.impl.Promise.$anonfun$transformWith$1(Promise.scala:37)           
        at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:60)                       
        at play.api.libs.streams.Execution$trampoline$.executeScheduled(Execution.scala:109)  
        at play.api.libs.streams.Execution$trampoline$.execute(Execution.scala:71)            
        at scala.concurrent.impl.CallbackRunnable.executeWithValue(Promise.scala:68)          
        at scala.concurrent.impl.Promise$DefaultPromise.dispatchOrAddCallback(Promise.scala:312)                                                                                             
        at scala.concurrent.impl.Promise$DefaultPromise.onComplete(Promise.scala:303)         
Caused by: java.time.format.DateTimeParseException: Text 'Mon, 26 Jul 1997 05:00:00' could not be parsed: Conflict found: Field DayOfWeek 6 differs from DayOfWeek 1 derived from 1997-07-26 
        at java.time.format.DateTimeFormatter.createError(DateTimeFormatter.java:1920)        
        at java.time.format.DateTimeFormatter.parse(DateTimeFormatter.java:1855)              
        at java.time.ZonedDateTime.parse(ZonedDateTime.java:597)                              
        at controllers.Assets$.parseModifiedDate(Assets.scala:575)                            
        at controllers.AssetsBuilder.$anonfun$maybeNotModified$4(Assets.scala:644)            
        at scala.Option.flatMap(Option.scala:171)                                             
        at controllers.AssetsBuilder.maybeNotModified(Assets.scala:643)                       
        at controllers.AssetsBuilder.$anonfun$assetAt$3(Assets.scala:744)                     
        at scala.concurrent.Future.$anonfun$flatMap$1(Future.scala:302)                       
        at scala.concurrent.impl.Promise.$anonfun$transformWith$1(Promise.scala:37)           
Caused by: java.time.DateTimeException: Conflict found: Field DayOfWeek 6 differs from DayOfWeek 1 derived from 1997-07-26                                                                   
        at java.time.format.Parsed.crossCheck(Parsed.java:647)                                
        at java.time.format.Parsed.crossCheck(Parsed.java:624)                                
        at java.time.format.Parsed.resolve(Parsed.java:246)                                   
        at java.time.format.DateTimeParseContext.toResolved(DateTimeParseContext.java:331)    
        at java.time.format.DateTimeFormatter.parseResolved0(DateTimeFormatter.java:1955)     
        at java.time.format.DateTimeFormatter.parse(DateTimeFormatter.java:1851)              
        at java.time.ZonedDateTime.parse(ZonedDateTime.java:597)                              
        at controllers.Assets$.parseModifiedDate(Assets.scala:575)                            
        at controllers.AssetsBuilder.$anonfun$maybeNotModified$4(Assets.scala:644)            
        at scala.Option.flatMap(Option.scala:171)
```

### Reproducible Test Case
```bash
curl 'http://localhost:9000/assets/test' -H 'If-Modified-Since: Mon, 26 Jul 1997 05:00:00 GMT' 
```

--- a/framework/src/play/src/main/scala/play/api/controllers/Assets.scala
+++ b/framework/src/play/src/main/scala/play/api/controllers/Assets.scala
@@ -30,6 +30,7 @@ package play.api.controllers {
 package controllers {
 
   import java.time._
+  import java.time.format.DateTimeParseException
   import javax.inject.Provider
 
   import akka.stream.scaladsl.StreamConverters
@@ -579,6 +580,9 @@ package controllers {
           }
         } catch {
           case e: IllegalArgumentException =>
+            Logger.debug(s"An invalid date was received: couldn't parse: $date", e)
+            None
+          case e: DateTimeParseException =>
             Logger.debug(s"An invalid date was received: couldn't parse: $date", e)
             None
         }

--- a/framework/src/play/src/test/scala/play/api/controllers/AssetsDateParsingSpec.scala
+++ b/framework/src/play/src/test/scala/play/api/controllers/AssetsDateParsingSpec.scala
@@ -37,6 +37,14 @@ class AssetsDateParsingSpec extends Specification {
       parseAndReformat("Wed Jan 07 2015 22:54:20 GMT-0800 (Pacific Standard Time)") must beSome("2015-01-08T06:54:20.000Z")
     }
 
+    "return None for improperly formatted date" in {
+      parseAndReformat("2015-01-07T05:00:00Z") must beNone
+    }
+
+    "return None for invalid date" in {
+      parseAndReformat("Mon, 26 Jul 1997 05:00:00 GMT") must beNone
+    }
+
     "not parse empty date header" in {
       parseAndReformat("") must beNone
     }

